module/localisation/BallFilter/src/BallFilter.cpp

```diff
@@ -35,53 +37,107 @@ namespace module::localisation {
 
-                // Apply exponential filter to rBCc
-                filtered_rBCc = cfg.smoothing_factor * rBCc + (1 - cfg.smoothing_factor) * filtered_rBCc;
-
-                // Generate message and emit
-                auto ball = std::make_unique<FilteredBall>();
-                Eigen::Affine3f Htc(sensors.Htw.cast<float>() * balls.Hcw.inverse().cast<float>());
-                Eigen::Affine3f Hrc(sensors.Hrw.cast<float>() * balls.Hcw.inverse().cast<float>());
-                ball->rBTt                = Htc * filtered_rBCc;
-                ball->rBRr                = Hrc * filtered_rBCc;
-                ball->rBCt                = Htc.rotation() * filtered_rBCc;
-                ball->time_of_measurement = NUClear::clock::now();
-
-                if (log_level <= NUClear::DEBUG) {
-                    log<NUClear::DEBUG>("rBTt: ", ball->rBTt.x(), ball->rBTt.y(), ball->rBTt.z());
-                    log<NUClear::DEBUG>("rBRr: ", ball->rBRr.x(), ball->rBRr.y(), ball->rBRr.z());
-                    log<NUClear::DEBUG>("rBCt: ", ball->rBCt.x(), ball->rBCt.y(), ball->rBCt.z());
-                    emit(graph("rBTt: ", ball->rBTt.x(), ball->rBTt.y(), ball->rBTt.z()));
-                    emit(graph("rBRr: ", ball->rBRr.x(), ball->rBRr.y(), ball->rBRr.z()));
-                    emit(graph("rBCt: ", ball->rBCt.x(), ball->rBCt.y(), ball->rBCt.z()));
-                }
+                    // Compute the time since the last update (in seconds)
+                    const auto dt = std::chrono::duration_cast<std::chrono::duration<float>>(NUClear::clock::now()
+                                                                                             - last_time_update)
+                                        .count();
+                    last_time_update = NUClear::clock::now();
+
+                    // Time update
+                    ukf.time(dt);
+
+                    // Measurement update
+                    ukf.measure(Eigen::Vector2f(rBWw.head<2>()),
+                                cfg.ukf.noise.measurement.position,
+                                MeasurementType::BALL_POSITION());
+
+                    // Get the new state, here we are assuming ball is on the ground
+                    state = BallModel<float>::StateVec(ukf.get_state());
+                    rBWw << state.rBWw.x(), state.rBWw.y(), fd.ball_radius;
+
+                    // Generate and emit message
+                    auto ball = std::make_unique<FilteredBall>();
+                    Eigen::Isometry3f Htw(sensors.Htw.cast<float>());
+                    Eigen::Isometry3f Hrw(sensors.Hrw.cast<float>());
+                    Eigen::Isometry3f Htc     = Eigen::Isometry3f(sensors.Htw.cast<float>()) * Hwc;
+                    ball->rBWw                = rBWw;
+                    ball->rBTt                = Htw * rBWw;
+                    ball->rBRr                = Hrw * rBWw;
+                    ball->rBCt                = Htc.rotation() * (Hcw * rBWw);
```

Comment:
Seems like overkill to compute all of these variations on `rBWw` when all of the transformations just come directly from `Sensors`. It would be better (and ultimately more accurate) yo just compute and store `rBWw`, whoever is triggering on your emitted message can then `With<Sensors>` if they need one of the other variations.

Reply:
While I very much agree with this, making this change is going to require a considerable amount of changes so I would prefer to leave updating this to a later PR. Also highlights the need to focus on getting the Director transition completed so we can phase out a lot of duplicate code and clean everything up.

Reply:
Just remember that these computed positions can (and probably will) be out of date when you use them, so there will likely be more error in them than there actually is.